### ISSUE TITLE
A guideline for beginners

### ISSUE THREAD

**Issue by Vayel** [crossbario/crossbarwww#26]
The WAMP ecosystem (WAMP, Crossbar.io and Autobahn) is a large project and it proves to be hard to know where to go for beginners. Indeed, there are several websites, several repositories, some examples on Crossbar.io website, some others on Autobahn websites... As a beginner, I deeply think that you should provide a guideline for people, a global view on the project and, as a beginner again, I would need: 

* 1. A starting point

I don't want to visit all the websites about the project to be able to start. As Crossbar.io is what connects all components, I suggest to use its website to welcome beginners.

* 2. A definition of WAMP

Just for people to know what it is. The slides 1, 2 and 3 from the WAMP website homepage would perfectly do the job.

* 2. A reason to get interested in WAMP

Give us an outlook on what WAMP allows us to do. No code here, just sexy demonstrations. But because most people won't understand why they should use it instead of AJAX or something else, that is the moment for explaining why WAMP was created and why it outstrips the others.

* 3. An introduction to WAMP and its concepts

I learnt why WAMP exists. Now, I want to understand the way it works. What should be taugh here is in the slidedeck on WAMP website homepage, which should be added on Crossbar.io website.

* 4. An overview of tools

Now, I want to use it. Explain to me how to install Crossbar.io and Autobahn before giving me some code I can tinker with. That's the step of experiments.

* 5. A complete example

The WAMP ecosystem looks great! But I now want to use it in a more serious project. So be it. Start a project with me from scratch and teach me how to configurate Crossbar.io, write my own components with Autobahn, code with the best practices...

Of course, that's just ideas.

Thanks!

**Comment by goeddea** [comment]
First of all - many thanks for this well-structured and detailed comment! This is the kind of feedback we're looking for.

Some follow-up questions:

* Do you see all of this as one single, long page on the landing page? Could it be something that is reachable via a prominent "Getting started" link?
* Have you played with the demos we provide - and which one(s) would you pick as being the best to introduce Crossbar.io? Any ideas for other demos that would convince you?

If you have more specific ideas, or additional comments regarding your experience as you get started, then they'd be very welcome.

And, regarding the documentation: This lives in the [Crossbar.io GitHub wiki](https://github.com/crossbario/crossbar/wiki), so feel free to edit it regarding mistakes you come across, or to add missing pieces! We know the documentation is very much a work in progress.

Again, thanks for this contribution, and I hope you continue to investigate what you can do with Crossbar.io!

**Comment by oberstet** [comment]
@Vayel this is invaluable feedback! thanks a lot. what you outline is a quick and clear path for beginners into the ecosystem. as soon as you are deep inside, such things become hard to see, because everything fits together and it's "just obvious". which, well, isn't of course;)

**Comment by Vayel** [comment]
Thanks a lot for your really-positive reactions!

@goeddea To me, the last two points should not live on the homepage. No code here. But I'm not sure of that.

Concerning the demos, I'll get interested in them more deeply before answering your question.

Besides, maybe the Crossbar.io website is not the right place to host it. See #27. If crossbar.io is still used to welcome beginners, the beginning og the introduction could be something like that (inspired by the slidedeck on wamp.ws homepage):

____

Crossbar.io is a router for WAMP, the Web Application Messaging Protocol. That protocol relies on websockets to enable different technologies, processes and machines communicate with each others in real time. Crossbar.io routes the messages from one component to another(s).

<img src="http://crossbar.io/static/img/gen/web_draft.png"></img>

Each component, written with AutobahnX, thus lives on its own, independent of the others. Crossbar.io is what connects them. WAMP, Crossbar.io and Autobahn constitute the WAMP ecosystem.

<img src="http://tavendo.com/static/img/gen/tavendo-technology-stack/tavendo-technology-stack.svg" alt="The WAMP ecosystem"></img>

Those tools allow you to do many great things!

DEMOS

Perhaps you wonder why you should use WAMP instead of AJAX or... 

INTRODUCTION TO BASIC WAMP CONCEPTS
______

But you're likely to do something far better (anyway, Sam from @sametmax would).

**Comment by Vayel** [comment]
Tried [here](http://vayel.github.io/crossbardocs/). To be continued.

**Comment by goeddea** [comment]
Wow! Will work on your suggestion during the coming days.

My problem with placing the demos this high up is that they're mostly very old, and almost exclusively illustrate PubSub (since this was the feature that WebMQ, a predecessor or Crossbar.io started out with back in the days of WAMP v1). I'd possibly only link to the votes demo + one other. 

What would you think of an inline demo, e.g. a mini-version of the votes? Not complicated to do,    and it would reduce friction in that the user then only needs to open one additional page to see it all in play. Possibly we could even have a version where the user can launch an in-browser backend for this, and have everything but Crossbar.io run locally in the browser.

**Comment by Vayel** [comment]
Indeed, the votes demo seems to do the job. It uses both RPC and PubSub and the code is quite clear and rather up-to-date. However, the backend appears to be currently useless and could mislead us about the fact that clients don't need any server to communicate (Crossbar.io does the job) with one another. So, perhaps you'd better give it a real part, like storing something in a database.

In all cases, the code should not be shown with the demo (see the first post). And I wonder if it's not Autobahn's part to introduce to it.

A substantial problem is that Crossbar.io is concerned by both Autobahn and WAMP. But repeating things between Web sites would make it more difficult for beginners to know where to learn. Each Web site should have a specific role and the others should only refer to it. Roles should be clearly defined not to have three Web sites about the WAMP ecosystem but about Crossbar.io, Autobahn and WAMP separately. Do you think `crossbar.io` should introduce to WAMP and Autobahn?

Would you mind developing your last sentence?

**Comment by goeddea** [comment]
Just a couple of things:

- I do think that the Crossbar.io website should give an introduction to WAMP concepts. We do not want a visitor to have to go to a different website in order to understand what Crossbar does. The introduction shouldn't be deep, but sufficient to get the basics. I agree with you that currently trying to explain things without mentioning WAMP is probably wrong. 
- Code should be there to show that CB/WAMP is simple to use. Words are nice, but code can't lie. If the code looks clean and easy to understand, then this convinces me much more than anything else. Note that we don't limit the code to Autobahn - the PHP example is for Thruway, and we're open to integrating examples from other languages/client libraries.
- My last sentence referred to the fact that you can run backend code in the browser - see https://github.com/crossbario/crossbarexamples/tree/master/votes/browser. A user could launch a backend in one browser tab, and then run clients in additional tabs or in additional browsers/on different devices. This would show that with WAMP you're really free to implement functionality where you want it. The downside would be that then launching the demo takes longer than with just two tabs  publishing and calling each other.

**Comment by Vayel** [comment]
Thanks for such a developped answer. What we need is a basic introduction to 
WAMP shared between `wamp.ws`, `crossbar.io` and `autobahnX.ws`: 

**1. A definition of WAMP**

* Websockets
* Communication
* Realtime
* Several machines
* Several technologies

**2. Sexy demonstration(s)**

* The votes (or another one, using both RPC and PubSub)
* Online (nothing to install)
* An overview of the code (see Alexander's great arguments above)
* With components written in several languages (JS and Python would be sufficient 
enough)

**3. An introduction to the basic concepts**

* Router (more details on `crossbar.io`)
* Components (more details on `autobahn.ws`, among others)
* Messages (more details on `wamp.ws`)
    * RPC
    * PubSub

Then, `crossbar.io` would: 

* teach how to configurate the router
* host tutorials involving the whole ecosystem (as the one about Django written 
by @sametmax)

`autobahn.ws` would:

* teach how to write components
* host the API Reference

`wamp.ws` would: 

* host the specification
* explain why WAMP is better than AJAX or anything else

Of course, each of them should clearly explain how to contribute.

A few questions:

* Should `crossbar.io` teach us how to write basic components as it does 
currently? 

To me, it shouldn't. For three reasons: the components are 
dependent on the technology (Autobahn is not mandatory) ; it wouldn't be 
clear enough for beginners to know where to start (especially if the code is 
different from one Web site to another) ; it would be more difficult for you 
to maintain things up-to-date.

Then, the last two points of the opening post would be dedicated to `autobahn.ws`. 

* Should the third point be illustrated with some code?

Currently, it is on the `wamp.ws` slidedeck. But the demos would already come with 
a little code. Thus, it seems not to be really useful to add some. Especially that 
it would cause the same problems as the ones raised in the precedent question.

__

Of course, these are just suggestions. I don't want to impose anything.

**Comment by goeddea** [comment]
Good arguments. Will get to this in more detail later. Just one thing:

Crossbar.io, WAMP and the Autobahn libraries are our projects, but the WAMP ecosystem is bigger than that. We can't just think about Autobahn, but want and need to include other clients. This is one reason for having one-page instructions for components in various languages & using a specific WAMP client library. I agree that it makes most sense to have additional information with the documentation of the respective library, e.g. the Crossbar.io docu should not comprise the AutobahnJS docu.

**Comment by Vayel** [comment]
> Crossbar.io, WAMP and the Autobahn libraries are our projects, but the WAMP ecosystem is bigger than that.

I agree for Crossbar.io and Autobahn, but WAMP (the protocol itself) is universal, isn't it?

> This is one reason for having one-page instructions for components in various languages & using a specific WAMP client library. I agree that it makes most sense to have additional information with the documentation of the respective library, e.g. the Crossbar.io docu should not comprise the AutobahnJS docu.

I did not understand because those sentences seem to contradict. Should Crossbar.io Web site teach how to write components with various librairies? Refering to the documentation of all librairies would be sufficient, wouldn't it?
